gluten-data/src/main/java/io/glutenproject/memory/nmm/NativeMemoryManagers.java

```diff
@@ -71,7 +72,21 @@ private static NativeMemoryManager createNativeMemoryManager(String name, Spille
                 new GlutenMemoryConsumer(
                     name,
                     taskMemoryManager,
-                    spiller,
+                    // call memory manager's shrink API, if no good then call the spiller
+                    Spillers.withOrder(
+                        (size, trigger) ->
+                            Optional.of(out.get())
+                                .map(nmm -> nmm.shrink(size))
+                                .orElseThrow(
+                                    () ->
+                                        new IllegalStateException(
+                                            ""
+                                                + "Shrink is requested before native "
+                                                + "memory manager is created. Try moving any "
+                                                + "actions about memory allocation out "
+                                                + "from the memory manager constructor.")),
+                        spiller // the input spiller, called after nmm.shrink was called
+                        ),
                     new GlutenMemoryConsumer.StatsBuilder() {
```

Comment:
Nice design! but we can decrease complexity by using spark's spill mechanism.
```
public long spill(long size, MemoryConsumer trigger) {
    if (trigger != this) {
        // shrink
    } else {
        // spill
    }
}
```

Reply:
This would introduce bug. Since self-spilling always happens in the case of Velox's plan execution.

Reply:
Let's use c1 for consumer1 and c2 for consumer2.

c1 will trigger other to spill then spill itself, if c1 has unused reservations when trigger c2 to spill, c2 will shrink then spill if shrink is not enough right? in this case, actual spill happens when there still has enough memory.

we can talk this offline.